spring-rabbit-test/src/main/java/org/springframework/amqp/rabbit/test/RabbitListenerTestHarness.java

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.amqp.rabbit.test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.mockito.Mockito;
+
+import org.springframework.amqp.rabbit.annotation.RabbitHandler;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.annotation.RabbitListenerAnnotationBeanPostProcessor;
+import org.springframework.amqp.rabbit.listener.MethodRabbitListenerEndpoint;
+import org.springframework.aop.framework.ProxyFactoryBean;
+import org.springframework.core.annotation.AnnotationAttributes;
+import org.springframework.core.annotation.AnnotationUtils;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * BeanPostProcessor extending {@link RabbitListenerAnnotationBeanPostProcessor}.
+ * Wraps the listener bean in a CGLIB proxy with an advice to capture the arguments
+ * and result (if any) in a blocking queue. Test cases can access the results
+ * by autowiring the test harness into test cases.
+ *
+ * @author Gary Russell
+ * @since 1.6
+ *
+ */
+public class RabbitListenerTestHarness extends RabbitListenerAnnotationBeanPostProcessor {
+
+	private final Log logger = LogFactory.getLog(this.getClass());
+
+	private final Map<String, CaptureAdvice> listenerCapture = new HashMap<String, CaptureAdvice>();
+
+	private final AnnotationAttributes attributes;
+
+	private final Map<String, Object> listeners = new HashMap<String, Object>();
+
+	public RabbitListenerTestHarness(AnnotationMetadata importMetadata) {
+		Map<String, Object> map = importMetadata.getAnnotationAttributes(RabbitListenerTest.class.getName());
+		this.attributes = AnnotationAttributes.fromMap(map);
+		Assert.notNull(this.attributes,
+				"@RabbitListenerTest is not present on importing class " + importMetadata.getClassName());
+	}
+
+	@Override
+	protected void processListener(MethodRabbitListenerEndpoint endpoint, RabbitListener rabbitListener, Object bean,
+			Object adminTarget, String beanName) {
+		String id = rabbitListener.id();
+		if (StringUtils.hasText(id)) {
+			if (this.attributes.getBoolean("spy")) {
+				Object spy = Mockito.spy(bean);
+				this.listeners.put(id, spy);
+				super.processListener(endpoint, rabbitListener, spy, adminTarget, beanName);
+				return;
+			}
+			else {
```

Comment:
I would like to have them together, not `if...else`.
The `spy` option assumes that we add or not the `.spy()` around our listeners.

Right the `ProxyFactoryBean` and `InvocationData` can be achieved with the `Spy` (somehow) as well, but why just don't expose it as is right now and wrap to `spy` if that?